numba/np/arrayobj.py

```diff
@@ -4523,6 +4523,45 @@ def impl(N, M=None, k=0, dtype=float):
     return impl
 
 
+@overload(np.diag_indices)
+def np_diag_indices(n, ndim=2):
+    if not isinstance(n, types.Integer):
+        msg = 'The argument "n" must be an integer'
+        raise errors.TypingError(msg)
+
+    if not isinstance(ndim, (int, types.Integer)):
+        msg = 'The argument "ndim" must be an integer'
+        raise errors.TypingError(msg)
+
+    def impl(n, ndim=2):
+        res = np.arange(n * ndim)
+        for i in range(n * ndim):
+            res[i] = res[i] % n
+        return res.reshape((ndim, n))
```

Comment:
Numba has a function called `tuple_setitem` from `cpython/unsafe/tuple.py` where you can use to set items in a tuple at runtime.

```suggestion
    # tup_init is used only to get the correct tuple type
    tup_init = (np.arange(1),) * 2
    from numba.cpython.unsafe.tuple import tuple_setitem

    def impl(n, ndim=2):
        res = np.arange(n * ndim)
        for i in range(n * ndim):
            res[i] = res[i] % n
        x = res.reshape((ndim, n))

        tup = tup_init
        for i in range(n):
            tup = tuple_setitem(tup, i, x[i])
        return tup
```

Reply:
Hi @guilhermeleobas, thank you for your suggestion. The only issue with this implementation is that the array elements that are being returned are immutable, which is different to what the NumPy function `np.diagflat` returns:
```python
import numpy as np
from numba import njit
@njit
def foodiagindices(n, ndim=2):
    res = np.diag_indices(n, ndim)
    return res
a=np.diag_indices(5)
print(a)
a[0][0] = 5
print(a)
c=foodiagindices(5)
print(c)
c[0][0] = 5
```
Returns

> (array([0, 1, 2, 3, 4]), array([0, 1, 2, 3, 4]))
(array([5, 1, 2, 3, 4]), array([5, 1, 2, 3, 4]))
(array([0, 1, 2, 3, 4]), array([0, 1, 2, 3, 4]))
Traceback (most recent call last):
  File "/home/kristian/Desktop/numba/a.py", line 13, in <module>
    c[0][0] = 5
ValueError: assignment destination is read-only

Another issue is that your proposed implementation implicitly assumes that `ndim` is equal to 2 at runtime by setting ` tup_init = (np.arange(1),) * 2`. Hence, the following code:
```python
import numpy as np
from numba import njit
@njit
def foodiagindices(n, ndim=2):
    res = np.diag_indices(n, ndim)
    return res
c = foodiagindices(5, 3)
```
throws a segmentation fault.